piet-coregraphics/src/text.rs

```diff
@@ -0,0 +1,433 @@
+    }
+
+    fn new_text_layout(
+        &mut self,
+        font: &Self::Font,
+        text: &str,
+        width: impl Into<Option<f64>>,
+    ) -> Self::TextLayoutBuilder {
+        let width_constraint = width.into().unwrap_or(f64::INFINITY);
+        let layout = CoreGraphicsTextLayout::new(font, text, width_constraint);
+        CoreGraphicsTextLayoutBuilder(layout)
+    }
+}
+
+impl Font for CoreGraphicsFont {}
+
+impl FontBuilder for CoreGraphicsFontBuilder {
+    type Out = CoreGraphicsFont;
+
+    fn build(self) -> Result<Self::Out, Error> {
+        //FIXME: this should be a 'missing font error'
+        self.0
+            .map(CoreGraphicsFont)
+            .ok_or_else(|| piet::new_error(piet::ErrorKind::MissingFeature))
+    }
+}
+
+impl TextLayoutBuilder for CoreGraphicsTextLayoutBuilder {
+    type Out = CoreGraphicsTextLayout;
+
+    fn build(self) -> Result<Self::Out, Error> {
+        Ok(self.0)
+    }
+}
+
+impl TextLayout for CoreGraphicsTextLayout {
+    fn width(&self) -> f64 {
+        self.frame_size.width
+    }
+
+    #[allow(clippy::float_cmp)]
+    fn update_width(&mut self, new_width: impl Into<Option<f64>>) -> Result<(), Error> {
+        let width = new_width.into().unwrap_or(f64::INFINITY);
+        if width.ceil() != self.width_constraint.ceil() {
+            let constraints = CGSize::new(width as CGFloat, CGFloat::INFINITY);
+            let char_range = self.attr_string.range();
+            let (frame_size, _) = self.framesetter.suggest_frame_size(char_range, constraints);
+            let rect = CGRect::new(&CGPoint::new(0.0, 0.0), &frame_size);
+            let path = CGPath::from_rect(rect, None);
+            self.width_constraint = width;
+            let frame = self.framesetter.create_frame(char_range, &path);
+            let line_count = frame.get_lines().len();
+            let line_origins = frame.get_line_origins(CFRange::init(0, line_count));
+            self.line_y_positions = line_origins
+                .iter()
+                .map(|l| frame_size.height - l.y)
+                .collect();
+            self.frame = Some(frame);
+            self.frame_size = Size::new(frame_size.width, frame_size.height);
+            self.rebuild_line_offsets();
+        }
+        Ok(())
+    }
+
+    fn line_text(&self, line_number: usize) -> Option<&str> {
+        self.line_range(line_number)
+            .map(|(start, end)| unsafe { self.string.get_unchecked(start..end) })
+    }
+
+    fn line_metric(&self, line_number: usize) -> Option<LineMetric> {
+        let lines = self.unwrap_frame().get_lines();
+        let line = lines.get(line_number.min(isize::max_value() as usize) as isize)?;
+        let line = Line::new(&line);
+        let typo_bounds = line.get_typographic_bounds();
+        let (start_offset, end_offset) = self.line_range(line_number)?;
+        let text = self.line_text(line_number)?;
+        //FIXME: this is just ascii whitespace
```

Comment:
What　could possibly go wrong?

Reply:
actually I didn't notice that apparently rust has at least some of the unicode tables? I mean: there's a `char::is_whitespace` method I can use here.

Reply:
Yeah, what Rust has is really limited, mostly just some case stuff and this. I think consensus is that real Unicode logic belongs in crates rather than core language.